Better exception handling/throwing

berdon opened the pull request:
Adding better exception handling/throwing

- Adding support for MorseL server and client options to
    - Throw exceptions on missing hub method invocation
    - Throw exceptions on missing client method invocation
    - Throw exceptions on callback method (On(...)) invocation
    - Throw exceptions by default on invalid/mangled messages
    - Rethrow exceptions thrown during the receive loop on a caller context (for client Connections)
- Added some context and clarification in some places

Commit:
Fixing compilation with dotnet 2.0, better socket closing check

Commit:
Adding better exception handling/throwing

- Adding support for MorseL server and client options to
  - Throw exceptions on missing hub method invocation
  - Throw exceptions on missing client method invocation
  - Throw exceptions on callback method (On(...)) invocation
  - Throw exceptions by default on invalid/mangled messages
  - Rethrow exceptions thrown during the receive loop on a caller context (for client Connections)
- Added some context and clarification in some places

retslig commented:
Looks pretty good.

ryanwersal commented on src/MorseL.Client/Connection.cs:
```diff
@@ -40,28 +62,39 @@ public class Connection
+            // Used to track exceptions and rethrow when able
+            Error += (exception) =>
+            {
+                if (_options.RethrowUnobservedExceptions)
```
I like the naming of this - it definitely implies the behavior you described.

ryanwersal commented on src/MorseL.Client/Connection.cs:
```diff
@@ -70,55 +103,136 @@ public Task StartAsync()
                             if (invocationDescriptor == null)
                             {
-                                _logger?.LogDebug("Invocation request for unknown hub method");
+                                // Valid JSON but unparseable into a known, typed invocation descriptor (unknown method name, invalid parameters)
```
Is there a way to tell the difference in a relatively trivial manner?

berdon replied:
Yeah...maybe... http://www.newtonsoft.com/json/help/html/JsonSchema.htm

berdon replied:
Are you ok with me creating a ticket to simplify the schema validation so I can push these changes?

ryanwersal replied:
Go for it. Truthfully the preview package is my only real concern but I'm okay with that as well. :ship:

ryanwersal commented on src/MorseL.Sockets/MorseL.Sockets.csproj:
```diff
@@ -15,6 +15,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Authorization" Version="1.1.1" />
     <PackageReference Include="Microsoft.AspNetCore.WebSockets" Version="1.0.1" />
+    <PackageReference Include="Nito.AsyncEx.Tasks" Version="5.0.0-pre-02" />
```
Is the preview package necessary?

berdon replied:
Necessary for .NET Standard 1.6+

ryanwersal commented on src/MorseL/Extensions/MorseLOptions.cs:
```diff
@@ -0,0 +1,39 @@
+using System.Collections.Generic;
+using System.Text;
+
+namespace MorseL.Extensions
```
This seems an odd namespace for this class?

berdon replied:
This mirrors SignalR's namespace for their AppBuilder/IServiceProvider extensions. I do agree that MorseLOptions inside of MorseL.Extensions is strange. Perhaps ServiceExtensions?

ryanwersal commented on src/MorseL/HubWebSocketHandler.cs:
```diff
@@ -152,27 +157,30 @@ private static bool IsHubMethod(MethodInfo methodInfo)
-                return;
+                // Really strange and unlikely, valid JSON and was known but not found here
+                // Likely not possible
+                await HandleMissingReceivedInvocationDescriptor(connection, invocationDescriptor);
```
I definitely appreciate the explicit error handling here. Compared to the previous "just send a message" this is a substantial improvement.

ryanwersal reviewed (commented):
Looks overall good to me. Glad to have a lot of improved error handling throughout.

Commit:
Rethrowing MorseLExceptions in HTTP Middleware

Rationale
- Good for unit tests
- Don't break anything, exceptions here get logged
- Can be handled explicitly using a wrapping middleware

Long Term
- Potentially add another option in MorseLOptions to rethrow these and
default to true

ryanwersal commented:
LGTM

berdon commented:
**Unit Tests**
Total tests: 36. Passed: 34. Failed: 0. Skipped: 2.
